### PR TITLE
feat(be): extend backend interface with completion_fd()

### PR DESCRIPTION
**Pull request by a-malakar**
- feat(be/vfio): add support for completion_fd()

### CODE REVIEW

File: include/libxnvme_queue.h [hunk 1]
```diff
@@ -153,3 +153,14 @@ typedef void (*xnvme_queue_cb)(struct xnvme_cmd_ctx *ctx, void *opaque);
  */
 int
 xnvme_queue_set_cb(struct xnvme_queue *queue, xnvme_queue_cb cb, void *cb_arg);
+
+/**
+ * Get the completion event fd on the given ::xnvme_queue
+ *
+ * @param queue Pointer to the ::xnvme_queue to query for outstanding commands
+ *
+ * @return On success, completion event fd is returned. On error, 0 is returned e.g.
```
Comment (safl) [inline include/libxnvme_queue.h:162]: I think this should be: ``On success, an eventfd() file descriptor is returned. On error, -errno is returned``.

File: include/xnvme_be_nosys.h [hunk 2]
```diff
@@ -95,7 +98,8 @@ xnvme_be_nosys_dev_close(struct xnvme_dev *dev);
 	{                                                                                 \
 		.cmd_io = xnvme_be_nosys_queue_cmd_io, .poke = xnvme_be_nosys_queue_poke, \
 		.wait = xnvme_be_nosys_queue_wait, .init = xnvme_be_nosys_queue_init,     \
-		.term = xnvme_be_nosys_queue_term, .id = "ENOSYS",                        \
+		.term = xnvme_be_nosys_queue_term, .get_completion_fd = xnvme_be_nosys_queue_get_completion_fd, \
+		.id = "ENOSYS",								\
```
Comment (safl) [inline include/xnvme_be_nosys.h:102]: Try running ``make format-all`` it should clean up the formating.

File: lib/xnvme_be_vfio_async.c [hunk 3]
```diff
@@ -150,6 +150,34 @@ xnvme_be_vfio_async_cmd_io(struct xnvme_cmd_ctx *ctx, void *dbuf, size_t dbuf_nb
 	return -EINVAL;
 }
 
+int
+xnvme_be_vfio_queue_get_completion_fd(struct xnvme_queue *queue)
```
Comment (safl) [inline lib/xnvme_be_vfio_async.c:154]: @birkelund would this be the correct way of enabling eventfd() notification for completion-notification on the ioqpair? And can the ``q->id`` be utilized as done below?

File: lib/xnvme_be_vfio_dev.c [hunk 4]
```diff
@@ -179,8 +177,8 @@ xnvme_be_vfio_dev_open(struct xnvme_dev *dev)
 	ctrl = _vfio_cref_lookup(&dev->ident);
 	if (!ctrl) {
 		struct nvme_ctrl_opts ctrl_opts = {
-			.nsqr = nqueues,
-			.ncqr = nqueues,
+			.nsqr = XNVME_BE_VFIO_NQUEUES_MAX - 1,
```
Comment (safl) [inline lib/xnvme_be_vfio_dev.c:180]: lgtm.

File: lib/xnvme_be_vfio_dev.c [hunk 5]
```diff
@@ -197,6 +195,14 @@ xnvme_be_vfio_dev_open(struct xnvme_dev *dev)
 		state->ctrl = ctrl;
 		state->qidmap = -1 & ~0x1;
 
+		for(int i = 0; i < XNVME_BE_VFIO_NQUEUES_MAX; i++) {
```
Comment (safl) [inline lib/xnvme_be_vfio_dev.c:198]: lgtm

**Review by safl — COMMENTED** [review]
This is excellent @a-malakar ! A couple of comments inline and then it would be great with a test utilizing the API introduced here. I have poked @birkelund for his expertise on the use of the ``libvfn`` API.

File: include/libxnvme.h [hunk 6]
```diff
@@ -30,6 +30,7 @@ extern "C" {
 #include <sys/types.h>
 #include <sys/uio.h>
 #include <time.h>
+#include <sys/eventfd.h>
```
Comment (safl) [inline include/libxnvme.h:33]: You can move this include into ``xnvme_be_vfio_async.c`` where ``eventfd()`` is called.

File: lib/xnvme_be_vfio_dev.c (hunk 4 above)
Comment (a-malakar) [reply lib/xnvme_be_vfio_dev.c:180]: Okay, lets go with the dynamic allocation approach.

File: include/xnvme_be_vfio.h [hunk 7]
```diff
@@ -12,14 +12,17 @@
...
+
 struct xnvme_be_vfio_state {
 	struct nvme_ctrl *ctrl;
 	unsigned long long qidmap; // Queue identifier bit map
 
 	struct nvme_sq *sq_sync; // Submission queue for synchronous IOs
 	struct nvme_cq *cq_sync; // Completion queue for synchronous IOs
 
-	uint8_t _rsvd[94];
+	int efds[XNVME_BE_VFIO_NQUEUES_MAX]; // completion event FDs
```
Comment (birkelund) [inline include/xnvme_be_vfio.h:24]: `vfio_set_irq()` is used to associate eventfds with a range of *vectors* - not queues. Many devices have fewer vectors available than the number of queues they support (i.e., the Optane I have at hand only has 9 MSI-X vectors).

You can get this number from the ``ctrl.pci.dev.irq_info``.

File: lib/xnvme_be_vfio_async.c (hunk 3 above)
Comment (birkelund) [reply lib/xnvme_be_vfio_async.c:155]: See above. efds should be sized according to number of vectors supported by the device. Not according to the number of queues supported.

**Review by birkelund — APPROVED** [review]

**Comment by karlowich** [comment]
Hi @a-malakar, I have refactored your example to allow us to choose a specific backend from the CLI.
With this functionality, I have added it to the CI.

@safl This looks good to me, can you take a look?

**Review by safl — APPROVED** [review]
Thanks @a-malakar , this is excellent! And thanks @karlowich !

**Pull request merged**